File: LMstudioConnection/LocalModel.js
const { LMStudioClient } = require("@lmstudio/sdk");
const fs = require('fs');
const readline = require('readline');
const nlp = require('compromise');  // Import compromise
const Tokenizer = require('nlp-tokenizer');
const stopword = require('stopword');  // Import the stopword module
var natural = require("natural"); 

// Setup reading line interface
const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout
});

async function main() {
  try {
      const client = new LMStudioClient();
      const model = await client.llm.load("lmstudio-community/Meta-Llama-3-8B-Instruct-GGUF/Meta-Llama-3-8B-Instruct-Q5_K_M.gguf");
      
      console.log("Reading the RawData.json file");
      // Read the JSON data
      const jsonData = JSON.parse(fs.readFileSync('C:\\Users\\Afro\\Projects\\JavascriptLMstudioTemplate\\LMstudioConnection\\RawData.json', 'utf8'));
  
      // Apply NLP processing using compromise
      const doc = nlp(jsonData.content);
      doc.normalize(); // normalize the text
  
      // Ensure we're passing a string to the tokenizer
      let rawText = doc.text() || ''; // Ensure there's a default empty string if nothing is returned
      //console.log("rawText"+ rawText);
      const tokenizer = new natural.WordTokenizer(); // Instantiate the tokenizer

      // rawText = rawText.replace(/,/g, '');
     // console.log("tokenizer" + tokenizer.tokenize(rawText))
      let tokens = tokenizer.tokenize(rawText)

    
      // Remove stopwords, ensuring all tokens are valid strings
      const filteredTokens = stopword.removeStopwords(tokens); // Use the stopword module to filter tokens
  
  
      // Rebuild the text with cleaned tokens
      const systemContent = filteredTokens.join('').toLowerCase();
      // TESTING PURPOconsole.log("SYS prompt"+systemContent);

  

    // Generate summary from the JSON data
    console.log(`\x1b[31mWelcome to your
        ___  _____   _   _                    ______       _
      / _ \\|_   _| | \\ | |                   | ___ \\     | |
     / /_\\ \\ | |   |  \\| | _____      _____  | |_/ / ___ | |_
     |  _  | | |   | . \` |/ _ \\ \\ /\\ / / __| | ___ \\/ _ \\| __|
     | | | |_| |_  | |\\  |  __/\\ V  V /\\__ \\ | |_/ / (_) | |_
     \\_| |_/\\___/  \\_| \\_/\\___| \\_/\\_/ |___/ \\____/ \\___/\\____/ By Jose Raul Valois\x1b[0m`);
             console.log()
             console.log(`Number of characters in the summarized content: ${systemContent.length}`);
      console.log(`Waiting on response from LMstudio`);

      const prediction = await model.respond([
          { role: "system", content: systemContent },
          { role: "user", content: "You are an AI News Bot Designed to help me digest large chunks of information in a few major key points. Identify major news events as key events in this format.EXAMPLE 'The text provides updates on various global events, including:[summarized key points of data]+[Cited Source]+[Date]'" }
      ]);

  
      if (prediction && prediction.content) {
        console.log(`\x1b[31mAI response content: ${prediction.content}\x1b[0m`);
      } else {
        console.error('Error: No content found in the AI response.');
      }
  
      console.log('Type "exit" to stop or ask another question related to the news articles.');
  
       // Continue to handle user inputs
          rl.on('line', async (input) => {
              if (input.toLowerCase() === 'exit') {
                  console.log('Exiting the AI assistant...');
                  rl.close();
              } else {
                  try {
                      const followup = await model.respond([
                          { role: "system", content: systemContent }, // Ensure content here matches expected structure
                          { role: "user", content: input }
                      ]);
                      
                      if (followup && followup.content) {
                          console.log(`\x1b[31mAI follow-up response: ${followup.content}\x1b[0m`); // Adjust per actual response structure
                      } else {
                          console.error('Follow-up error: No content found in the AI response.');
                      }
                      console.log('Type "exit" to stop or ask another question.');
                  } catch (error) {
                      console.error('Error handling follow-up:', error);
                  }
              }
          });
      } catch (error) {
          console.error('Error:', error);
          rl.close();
      }
  }
  
  main();

File: LlamafileConnection/Localmodel.py
import json
import re
import asyncio
from aioconsole import ainput
import aiohttp
import spacy

BASE_URL = "http://127.0.0.1:8080/v1/chat/completions"

def sanitize_string(string):
    return re.sub(r'[^\w\s.,;!?\'"]', '', string)

async def generate_summary(session, system_content, user_input=None):
    

# Load the spaCy English model
   nlp = spacy.load("en_core_web_sm")

# Reading the JSON data
   with open('C:\\Users\\Afro\\Projects\\JavascriptLMstudioTemplate\\LMstudioConnection\\RawData.json', 'r', encoding='utf-8') as file:
    jsonData = json.load(file)

# Apply NLP processing
    doc = nlp(jsonData['content'])

 # Normalize the text and tokenize
    tokens = [token.text for token in doc if not token.is_punct and not token.is_space]
 
# Filter tokens: longer than 2 characters and purely alphabetic
    filtered_tokens = [token for token in tokens if len(token) > 2 and token.isalpha()]

# Rebuild the text from filtered tokens
    system_content = ' '.join(filtered_tokens)
    
   # print(system_content)
    print("Character count of cleaned data"+ len(system_content))
    messages = [
        {"role": "system", "content": system_content},
        {"role": "user", "content": "You are an AI News Bot Designed to help me digest large chunks of information in a few major key points. Summarize the text with the key points in this format. EXAMPLE 'The text provides updates on various global events, including:'"}
    ]
    if user_input:
        messages.append({"role": "user", "content": user_input})

    payload = {
        "model": "Meta-Llama-3-8B-Instruct-Q5_K_M.gguf",
        "stream": True,
        "messages": messages
    }
    headers = {"Content-Type": "application/json", "Accept": "text/event-stream"}

    full_response_content = []

    async with session.post(BASE_URL, json=payload, headers=headers) as response:
        try:
            if response.headers.get('Content-Type') == 'text/event-stream':
                async for line in response.content:
                    if line.strip():
                        json_data = json.loads(line.decode().split("data: ", 1)[1])  # Parsing the streamed JSON object
                        content = json_data.get("choices", [{}])[0].get("delta", {}).get("content", "")
                        full_response_content.append(content)
            return ''.join(full_response_content)  # Join all content parts into a single string
        except Exception as e:
            print(f"Unexpected error occurred while handling the stream: {e}")
            return None

async def main():
    async with aiohttp.ClientSession() as session:
        system_content = "Initial system content here"
        user_input = None
        print("Reading the Output.json file with your LLamafile server connection...")
        summary = await generate_summary(session, system_content, user_input)
        if summary:
            print(f"AI response content: {summary}")
        else:
            print('Error: No content found in the AI response.')

        while True:
            user_input = await ainput("Type 'exit' to stop or ask another question: ")
            if user_input.lower() == 'exit':
                print('Exiting the AI assistant...')
                break

            follow_up_response = await generate_summary(session, system_content, user_input)
            if follow_up_response:
                print(f"AI follow-up response: {follow_up_response}")
            else:
                print('Follow-up error: No content found in the AI response.')

if __name__ == '__main__':
    asyncio.run(main())